### ISSUE TITLE
How to deploy .Net core app types with IIS

### ISSUE THREAD

**Issue by bthomsons** [aspnet/Home#1651]
I'm trying to work out the best way of deploying a .net core web app. For legacy reasons the .net core application needs to be a child of another asp.net web application. This is something we can do in IIS. For example in IIS, we are set up like this

../App -> points to a standard asp.net app
../App/ChildApp -> should point to our .net core app

(this is because links in /App point to /App/ChildApp and should be under the same url)

I have managed to get the application to publish on build but when running on IIS you cannot do another publish unless you kill the ChildApp.exe process - obviously not a very good developer experience.

I wanted to ask a few things:

1) Can you deploy a dll instead of an exe like you can in standard asp.net (it seems to allude to the fact you can here https://docs.microsoft.com/en-us/dotnet/articles/core/app-types but I'm not clear which part of the project.json is actually doing this)?

2) Will deploying as a dll even stop the need to kill a process as I understand the following command is run "dotnet run ChildApp.dll"?

3) Is there some other way of achieving what I need without deploying as a dll

**Comment by GuardRex** [comment]
> kill the ChildApp.exe process

I take it you mean via `app_offline.htm` or by stopping/restarting the AppPool with PowerShell IIS Admin cmdlets ... or via Web Deploy. I mean you aren't actually manually ending the process in Task Manager, right?

1. Yes. To do that, you will publish a portable app type, which is determined by the `"type": "platform"` designation on the `Microsoft.NETCore.App` package. You can remove the `runtimes` section of `project.json` and publish with `dotnet publish -c Release`. The expected outputs can be seen here: https://docs.asp.net/en/latest/hosting/directory-structure.html

2. The deployed portable app will still have the DLL file locked when the AppPool and process is running. It doesn't solve the need to stop the AppPool for future deployments. I'm not sure what you mean by `dotnet run ChildApp.dll`. The app would be "run" by the ASP.NET Core Module (ANCM) using `dotnet .\ChildApp.dll` (i.e., executed only) The `dotnet run` command is a build-and-run command (i.e., compile and execute). https://docs.microsoft.com/en-us/dotnet/articles/core/tools/dotnet-run

3. If you're asking about the file lock ... no. You must use `app_offline.htm`, PowerShell, manually stop/start the AppPool, or use Web Deploy (MSDeploy) ... or some other means to release the file lock.

**Comment by bthomsons** [comment]
Hi @GuardRex, thanks for your reply. Restarting the app pool or IIS didn't kill the process, I actually had to kill it manually, which didn't seem right to me. I have re-created the project from scratch this morning and re-tested and this works as expected now after an IIS reset. Thank you for solving 2) and 3)

For 1) we have some .net 46 frameworks which we will need to include in the application. If I put the type:platform as follows

>   "dependencies": {
    "Microsoft.NETCore.App": {
      "version": "1.0.0",
      "type": "platform"
    },
    "Microsoft.AspNetCore.Diagnostics": "1.0.0",
    "Microsoft.AspNetCore.Server.IISIntegration": "1.0.0",
    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0",
    "Microsoft.Extensions.Logging.Console": "1.0.0"
  },
  "tools": {
    "Microsoft.AspNetCore.Server.IISIntegration.Tools": "1.0.0-preview2-final"
  },
  "frameworks": {
    "dnx46": { }
  },

In the project.json file I get an error that Microsoft.NETCore.App is not supported.

**Comment by GuardRex** [comment]
Yes, that's correct. The `Microsoft.NETCore.App` meta-package is only compatible with the  `netcoreapp1.0` TFM. You're targeting the .NET Framework, so you should drop that package. If you need BCL support in your app, use `"NETStandard.Library": "1.6.0"` ( https://www.nuget.org/packages/NETStandard.Library/ ) or pick-and-choose packages that support .NET Framework 4.6 (`.net46` TFM). ......and that's the other thing you'll need to change. DNX is :skull: dead and buried :skull:, so you'll need to change the TFM from `dnx46` to `net46`.

If you have further questions and would like to chat with devs, join the .NET Core Slack Channel: http://tattoocoder.com/aspnet-slack-sign-up/ ... there are usually dozens of devs in there who can answer questions and help troubleshoot problems. Join the #general channel, and there are specific channels for #aspnet-core, #dotnet-cli, #dotnet-core, #iis, #kestrel, #identity, #xunit, and #mvc (just to name a few!).

**Comment by bthomsons** [comment]
Thanks @GuardRex I guess that means that because we can't have `"NETStandard.Library": "1.6.0"` use the type: platform, that there's no way of having an app which targets the framework net46 to deploy to a dll instead of an exe. Is that correct? If it is, is there going to be support for this in the future?

**Comment by GuardRex** [comment]
Ah ... my bad. I'm 100% Core CLR here, and I don't work with the "ancient" .NET Framework any more. :smile:

Publish for `net461`, and you'll see the DLL in the published output.

**Comment by moozzyk** [comment]
If you target full .NET Framework you will get .exe. To stop your application before deploying a newer version drop a file called "app_offline.htm" to the app folder. IIS (AspNetCoreModule) monitors the folder and if the file is present it will 
- drain requests (by sending Ctrl+C/Break signal to your app)
- stop your application, thus unblocking the files
- serve the contents of the app_offline.htm to the user
Once you finish deploying a new version of your application delete the app_offline.htm and IIS (AspNetCoreModule) will start your application.

**Comment by GuardRex** [comment]
> If you target full .NET Framework you will get .exe.

It appears that when `"emitEntryPoint"` is set to `true` the `.exe` appears in in the output. I checked with `net46`, and this was the behavior. [I really **should** take some time and get more familiar with .NET Framework apps out of `dotent cli` tooling, but it feels like that's :boat: against the wind. :smile:]

**Comment by gforceg** [comment]
@GuardRex - 

I tried to join the slack group but my employer's firewall has blocked access to slack. :stop_sign: 

I have a very simple dotnet core application (of the project.csproj variety, not the project.json variety) that I am trying to deploy to IIS using the 'WinRM - IIS Web App Deployment Task' in a Release (TFS 2017).

Upon the first release, everything is successful except that IIS shows my application as a folder. If I go to IIS Manager, expand 'Default Web Site' and right click my application's folder, I can choose 'Convert to Application', and everything works fine. All of my subsequent releases work fine after that. :smile: 

Do you know what steps I need to take to get IIS to recognize my initial Release as an application rather than simply a folder? Sorry I'm not an IIS / dotnet guy and I haven't read an answer that I understand yet.

Thanks.

**Comment by GuardRex** [comment]
@gforceg Idk (I'm 100% using PS these days for deployment). Let's ping @vijayrkn to see if he has a tip or knows who to ask.

**Comment by gforceg** [comment]
@GuardRex Thanks for responding so quickly. I know you're crazy busy.

BTW when I choose 'convert to application' the first go around, I am selecting a separate app pool for my application was is created by my TFS 2017 Release.

**Comment by gforceg** [comment]
I was able to get this working in powershell! *YAY!*

The only problem is that I cannot set the dotnet runtime to 'no managed code' without logging into IIS.

I have a reference to the AppPool in powershell but when I set `$appPool.managedRuntimeVersion = $null`, IIS still shows it as set to 'V4.0' though the variable $appPool in powershell shows the value as null. I suspect IIS is correct.

**Comment by gforceg** [comment]
It works! I had to use Set-ItemProperty rather than trying to modify the apppool  object directly.

**Comment by aspnet-hello** [comment]
This issue is being closed because it has not been updated in 3 months.

We apologize if this causes any inconvenience. We ask that if you are still encountering this issue, please log a new issue with updated information and we will investigate.
